### ISSUE TITLE
Setting per-address max TTL

### ISSUE THREAD

**Issue by lulf** [EnMasseProject/enmasse#3158]
**Overview**

Allow an address/address plan for queues to be configured with a max TTL for messages.

**Design**

The `addressplan` and `address` crds will be extended to carry a TTL specification in the form of a minimum and maximum TTL values.   The maximum and minimum values will clamp the TTL value of an incoming messages to that address.  The clamp provided by an address may only narrow the clamp provided by the plan, it cannot cause it to be widened.   

1. If a messages arrives with a TTL value that is greater than the maximum allowed by the clamp, its TTL is changed to the maximum.
1. If a messages arrives with a TTL value that is less than the minumum allowed by the clamp, its TTL is changed to the minimum.
1. Messages that arrive 'without TTL' are considered to have a TTL of infinity.
1. A unspecified TTL minimum (at plan or address) is considered to have the value 0.
1. A unspecified TTL maximum (at plan or address) is consider to have the value infinity.

The TTL specification will look like this:

```
ttl:
  maximum: <duration in milliseconds or -1 (represents infinity)>
  minimum: <value in millseconds>
```

The same TTL specification will be used in in the spec sections of the `addressplan` and `address`.  The `status` section of the address will include the `ttl` section in order to reveal the TTL values that are in force for that address.

If a message expires, the behaviour at the Broker will be to drop the message.  That is, there will be no expiry address specified.  (Expiry address will be added later by shared).

Message expiry will apply to queue and topic address types. Message expiry values for cast addresses will be ignored.  It will apply only to newly incoming messages.

The feature will be implemented for broker and standard.

**Tasklist**

- [x] CRD extension
- [x] Extend agent to propagate max/min TTL from plan/address to the address-setting
- [x] Extend agent to update status section of the brokered addresses
- [x] Extend stanard controller to update the status section of the standard addresses.
- [x] Write system tests

**Out of scope**

- Metrics
- UI

**Comment by lulf** [comment]
@k-wall Should we allow for specifying maximum and minimum using more human-readable input? I.e. '6s', or '2m' ?

**Comment by k-wall** [comment]
> @k-wall Should we allow for specifying maximum and minimum using more human-readable input? I.e. '6s', or '2m' ?

I wondered that myself.  My initial instinct was for ISO 8601 durations but saw this wasn't supported by Golang's core.

I looked around for a common kubernetes or openshift CRD conventions but drew a blank.   For readiness/liveness probes they use a second value with the units embedded in the name (PeriodSeconds).   Go lang supports https://golang.org/pkg/time/#ParseDuration but I didn't see cases where CRDs had adopted the stringified form for their Kubernetes CRD fields.

There was another consideration too - the need to encode the infinity value, I'm using -1.  That might look odd expressed -1ms.. would -1s work too?  In the end I followed the AMQP convention and went for a millisecond value.
